### PR TITLE
Move name resolution logic to a dedicated file

### PR DESCRIPTION
**Pull request by cjgillot**
The code resolution logic from an Ident is scattered between several files.

The first commits creates `rustc_resolve::probe` module to hold the different mutually recursive functions together. Just a move, no code change.
The following commits attempt to make the logic a bit more readable.

The two fields `last_import_segment` and `unusable_binding` are replaced by function parameters.
In order to manage the fallout, `maybe_` variants of the function are added, dedicated to speculative resolution.

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
Some preliminary comments:
- It's not clear why "probe" was chosen as the name
- I'd prefer to move the `fn report_error` stuff to diagnostics.rs leaving other files primarily for "reference implementation of the language", if that makes sense
- It's unfortunate that multiple versions of functions like `fn maybe_foo` + `fn foo` + `fn foo_with_ribs` + etc have to be introduced as a replacement for having optional function parameters at language level. The "maybe" naming doesn't even reflect that we are simply omitting some parameters with default values.
I wish there was a better way to address this, but the alternative of creating a separate new `FooArgs` struct for every such function is even more awful.

**Comment by cjgillot** [comment]
I believe this version addresses your first round of reviews.  The refactor of `resolve_path` has been removed.

> Some preliminary comments:
> 
>     * It's not clear why "probe" was chosen as the name
> 

Just because I found `rustc_resolve::resolve` silly.  What would you prefer?

>     * I'd prefer to move the `fn report_error` stuff to diagnostics.rs leaving other files primarily for "reference implementation of the language", if that makes sense
> 

Done.

>     * It's unfortunate that multiple versions of functions like `fn maybe_foo` + `fn foo` + `fn foo_with_ribs` + etc have to be introduced as a replacement for having optional function parameters at language level. The "maybe" naming doesn't even reflect that we are simply omitting some parameters with default values.
>       I wish there was a better way to address this, but the alternative of creating a separate new `FooArgs` struct for every such function is even more awful.

I understand the `maybe_` version to be suitable for speculative resolution.  Is it correct with those parameters?

**Comment by petrochenkov** [comment]
>Just because I found `rustc_resolve::resolve` silly.  What would you prefer?

Well, if we have `macros.rs` for macro-specific stuff, and `imports.rs` for import-specific stuff, them maybe we should have `ident(s).rs` for identifier resolution stuff (which I see moved to this new file).

(`resolve_path(_with_ribs)` should ideally be a nearly trivial loop over resolving identifiers, so it doesn't deserve its own file or file naming.)

File: compiler/rustc_resolve/src/probe.rs [hunk 1]
```diff
@@ -851,12 +851,65 @@ impl<'a> Resolver<'a> {
...
                 if let NameBindingKind::Res(_, true) = binding.kind {
                     self.macro_expanded_macro_export_errors.insert((path_span, binding.span));
                 }
             }
+
+            // If the primary binding is unusable, search further and return the shadowed glob
+            // binding if it exists. What we really want here is having two separate scopes in
+            // a module - one for non-globs and one for globs, but until that's done use this
+            // hack to avoid inconsistent resolution ICEs during import validation.
+            if let Some(unusable_binding) = self.unusable_binding
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_resolve/src/probe.rs]: I'm not sure I understand what you mean.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_resolve/src/probe.rs]: @cjgillot 
https://github.com/petrochenkov/rust/commit/d72946b87c44a79d33d27308433f85c6c1e2478c
(Maybe there's no need to move it to a separate function.)

File: compiler/rustc_resolve/src/ident.rs [hunk 2]
```diff
@@ -0,0 +1,1590 @@
...
+use crate::late::{ConstantItemKind, HasGenericParams, PathSource, Rib, RibKind};
+use crate::macros::{sub_namespace_match, MacroRulesScope};
+use crate::{AmbiguityError, AmbiguityErrorMisc, AmbiguityKind, Determinacy, Finalize};
+use crate::{
+    ImportKind, LexicalScopeBinding, Module, ModuleKind, ModuleOrUniformRoot, NameBinding,
+};
+use crate::{
+    NameBindingKind, ParentScope, PathResult, PrivacyError, Res, ResolutionError, Resolver, Scope,
+    ScopeSet, Segment,
+};
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:25]: Nit: could you split imports to be single-line in all the changed files?

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/95405#discussion_r848690424 applied.

**Pull request merged**